### PR TITLE
Use BeanFactory's getBean in findBean

### PR DESCRIPTION
**Pull request by delegacy**
Motivation:

- `spring-boot-webflux-starter` app does not start when multiple
`MeterRegistry`s are registerted

```
org.springframework.context.ApplicationContextException: Unable to start reactive web server; nested exception is java.lang.IllegalStateException: Too many MeterRegistry beans: [io.micrometer.cloudwatch.CloudWatchMeterRegistry@6dd79214, io.micrometer.prometheus.PrometheusMeterRegistry@2c47a053, io.micrometer.core.instrument.composite.CompositeMeterRegistry@7767bd4e] (expected: 1)
	at org.springframework.boot.web.reactive.context.ReactiveWebServerApplicationContext.onRefresh(ReactiveWebServerApplicationContext.java:82)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:543)
	at org.springframework.boot.web.reactive.context.ReactiveWebServerApplicationContext.refresh(ReactiveWebServerApplicationContext.java:67)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:775)
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:397)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:316)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1260)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1248)
	at ...
Caused by: java.lang.IllegalStateException: Too many MeterRegistry beans: [io.micrometer.cloudwatch.CloudWatchMeterRegistry@6dd79214, io.micrometer.prometheus.PrometheusMeterRegistry@2c47a053, io.micrometer.core.instrument.composite.CompositeMeterRegistry@7767bd4e] (expected: 1)
	at com.linecorp.armeria.spring.web.reactive.ArmeriaReactiveWebServerFactory.findBean(ArmeriaReactiveWebServerFactory.java:268)
	at com.linecorp.armeria.spring.web.reactive.ArmeriaReactiveWebServerFactory.configureArmeriaService(ArmeriaReactiveWebServerFactory.java:242)
	at com.linecorp.armeria.spring.web.reactive.ArmeriaReactiveWebServerFactory.lambda$getWebServer$2(ArmeriaReactiveWebServerFactory.java:189)
	at java.base/java.util.Optional.ifPresent(Optional.java:183)
	at com.linecorp.armeria.spring.web.reactive.ArmeriaReactiveWebServerFactory.getWebServer(ArmeriaReactiveWebServerFactory.java:189)
	at org.springframework.boot.web.reactive.context.ReactiveWebServerApplicationContext$ServerManager.<init>(ReactiveWebServerApplicationContext.java:202)
	at org.springframework.boot.web.reactive.context.ReactiveWebServerApplicationContext$ServerManager.get(ReactiveWebServerApplicationContext.java:221)
	at org.springframework.boot.web.reactive.context.ReactiveWebServerApplicationContext.createWebServer(ReactiveWebServerApplicationContext.java:90)
	at org.springframework.boot.web.reactive.context.ReactiveWebServerApplicationContext.onRefresh(ReactiveWebServerApplicationContext.java:79)
	... 8 common frames omitted
```

Modification:

- Use BeanFactory's getBean when finding a single bean
  - BeanFactory's getBean resolves a primary bean among many of the same type

### CODE REVIEW

File: spring/boot-webflux-autoconfigure/src/main/java/com/linecorp/armeria/spring/web/reactive/ArmeriaReactiveWebServerFactory.java [hunk 1]
```diff
@@ -257,16 +260,14 @@ private void configureArmeriaService(ServerBuilder sb, ArmeriaSettings settings)
...
-            case 1:
-                return Optional.of(beans.get(0));
-            default:
-                throw new IllegalStateException("Too many " + clazz.getSimpleName() + " beans: " +
-                                                beans + " (expected: 1)");
+    @VisibleForTesting
+    <T> Optional<T> findBean(Class<T> clazz) {
+        try {
+            return Optional.of(beanFactory.getBean(clazz));
+        } catch (NoUniqueBeanDefinitionException e) {
```
Comment (hyangtack) [inline spring/boot-webflux-autoconfigure/src/main/java/com/linecorp/armeria/spring/web/reactive/ArmeriaReactiveWebServerFactory.java]: Just curious. From `getBean` javadoc:
```
@throws NoUniqueBeanDefinitionException if more than one bean of the given type was found
```
So I'm curious whether the problem you described would be resolved or not. Is this fix working when multiple `MeterRegistry` beans are specified? How about adding a test for it?

Comment (delegacy) [reply spring/boot-webflux-autoconfigure/src/main/java/com/linecorp/armeria/spring/web/reactive/ArmeriaReactiveWebServerFactory.java]: I checked with my personal project that it is working.

I thought about adding tests for this change, it seems like testing against Spring framework, so I just gave up.

Let me think what I can do.

Comment (delegacy) [reply spring/boot-webflux-autoconfigure/src/main/java/com/linecorp/armeria/spring/web/reactive/ArmeriaReactiveWebServerFactory.java]: Added tests to show behavioral difference.

File: spring/boot-webflux-autoconfigure/src/main/java/com/linecorp/armeria/spring/web/reactive/ArmeriaReactiveWebServerFactory.java [hunk 2]
```diff
@@ -258,15 +260,12 @@ private void configureArmeriaService(ServerBuilder sb, ArmeriaSettings settings)
...
-            case 0:
-                return Optional.empty();
-            case 1:
-                return Optional.of(beans.get(0));
-            default:
-                throw new IllegalStateException("Too many " + clazz.getSimpleName() + " beans: " +
-                                                beans + " (expected: 1)");
+        try {
+            return Optional.of(beanFactory.getBean(clazz));
+        } catch (NoUniqueBeanDefinitionException e) {
```
Comment (minwoox) [inline spring/boot-webflux-autoconfigure/src/main/java/com/linecorp/armeria/spring/web/reactive/ArmeriaReactiveWebServerFactory.java]: If this is finding the first one among beans, shouldn't we remove this clause and just leave catching `NoSuchBeanDefinitionException` clause?

**Comment by trustin** [comment]
Thank you, @delegacy !

**Comment by delegacy** [comment]
Thank you, all.

**Pull request merged**